### PR TITLE
♻️✨Register common video actions and events

### PR DESCRIPTION
**Pull request by alanorozco**
Partial for #13674

For comparison, se [`video-manager-impl.js`](https://github.com/ampproject/amphtml/blob/b3af26ce04c264495f231269273c5c78d8bf447a/src/service/video-manager-impl.js).

### CODE REVIEW

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 1]
```diff
@@ -58,32 +68,197 @@ export class VideoService {
   constructor(ampdoc) {
     /** @private @const {!../../../src/service/ampdoc-impl.AmpDoc} */
     this.ampdoc_ = ampdoc;
+
+    this.tick_ = new LazyObservable(() => {
+      const {win} = ampdoc;
+      const timer = Services.timerFor(win);
+
+      timer.delay(() => this.tick_.fire(), 1000);
```
Comment (nainar) [inline extensions/amp-video-service/0.1/amp-video-service.js]: Why is this delay here?

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 2]
```diff
@@ -58,32 +68,197 @@ export class VideoService {
...
     warnUnimplemented('Video analytics');
-    return Promise.resolve();
   }
 
   /**
    * @param {!AmpElement} unusedVideo
    * @param {!../../../src/observable.Observable<boolean>} unusedObservable
    */
-  delegateAutoplay(unusedVideo, unusedObservable) {
+  delegateAutoplay(unusedVideo , unusedObservable) {
```
Comment (nainar) [inline extensions/amp-video-service/0.1/amp-video-service.js]: nit: remove

File: extensions/amp-video-service/0.1/amp-video-service.js (hunk 1 above)
Comment (alanorozco) [reply extensions/amp-video-service/0.1/amp-video-service.js]: Should be every second. Fixed.

File: extensions/amp-video-service/0.1/amp-video-service.js (hunk 2 above)
Comment (alanorozco) [reply extensions/amp-video-service/0.1/amp-video-service.js]: Thanks, fixed!

**Comment by alanorozco** [comment]
@choumx 

Current master: 243982
This branch after merge: 243982
Delta: 0

:)

LazyObservable is only used in the extension, I believe that's the reason doesn't affect the size of v0.js at all.

**Comment by choumx** [comment]
Thanks for checking the binary size delta. For context, I'm planning to cap the size of v0.js soon so we have a fighting chance of completing the "reduce size by 50%" KR.

I'm not sure I understand the purpose of `LazyObservable`. Why not just use an `Observable` with a boolean in the handler?

```js
const o = new Observable();
o.add(() => {
  if (!this.flag_) {
    // Do something only once.
  }
  this.flag_ = true;
});
```

**Comment by alanorozco** [comment]
@choumx 

The concept behind `LazyObservable` is that the mechanism that fires it is "lazy", since it won't be started until a consumer observes it. This is at the firing level, not at the handler level.

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 3]
```diff
@@ -76,14 +140,157 @@ export class VideoService {
...
+    this.ampdoc_ = ampdoc;
+
+    /** @private @const{!../../../src/video-interface.VideoInterface} */
+    this.video_ = video;
+
+    /** @private @const {!../../../src/observable.ObservableInterface<void>} */
+    this.tick_ = tick;
+
+    // These functions are bound (instead of being arrow functions) so that
+    // they reference the prototype of the class, which is only created once.
```
Comment (choumx) [inline extensions/amp-video-service/0.1/amp-video-service.js]: Not sure what this means. Doesn't `this.onTick_ == VideoEntry.prototype.onTick_`?

**Comment by alanorozco** [comment]
@choumx 

- Removed `LazyObservable` in favor of scoped `onTick`/`onPlaybackTick`.
- Enabled/updated test that was failing with `LazyObservable`  (whoops).

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 4]
```diff
@@ -76,14 +148,159 @@ export class VideoService {
...
+      this.isPlaying = true;
+    });
+  }
+
+  /**
+   * Register common actions such as play, pause, etc... so they can be called
+   * using AMP Actions, e.g.: `<button on="tap:myVideo.play">`.
+   * @visibleForTesting
+   */
+  registerCommonActions() {
```
Comment (choumx) [inline extensions/amp-video-service/0.1/amp-video-service.js]: You can still mark these `@private` and access them normally in unit tests (not integration tests I believe) to allow CC renaming.

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 5]
```diff
@@ -76,14 +148,159 @@ export class VideoService {
...
+   * @param {!Observable<boolean>} unusedObservable
    */
   delegateAutoplay(unusedVideo, unusedObservable) {
     warnUnimplemented('Autoplay delegation');
   }
 }
 
 
+/** @visibleForTesting */
+export class VideoEntry {
```
Comment (choumx) [inline extensions/amp-video-service/0.1/amp-video-service.js]: Brief description of this class?

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 6]
```diff
@@ -76,14 +148,159 @@ export class VideoService {
...
+      if (!this.isPlaying) {
+        return;
+      }
+      handler();
+    });
+  }
+
+  /** */
+  install() {
+    const {element} = this.video_;
```
Comment (aghassemi) [inline extensions/amp-video-service/0.1/amp-video-service.js]: I can't think of cases where we don't want to call install after creating. Maybe `create` and call this automatically so we caller doesn't have to?

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 7]
```diff
@@ -76,14 +148,159 @@ export class VideoService {
...
+    /** @visibleForTesting {boolean} */
+    this.isPlaying = false;
+  }
+
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!VideoService} videoService
+   * @param {!../../../src/video-interface.VideoInterface} video
+   */
+  static create(ampdoc, videoService, video) {
```
Comment (aghassemi) [inline extensions/amp-video-service/0.1/amp-video-service.js]: I like this, we don't use this pattern often enough.

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 8]
```diff
@@ -76,14 +148,159 @@ export class VideoService {
...
+   * Triggers a LOW-TRUST timeupdate event consumable by AMP actions if
+   * required.
+   * Frequency of this event is controlled by VideoService.onTick() and is
+   * every second for now.
+   * @visibleForTesting
+   */
+  maybeTriggerTimeUpdate() {
+    const {element} = this.video_;
+
+    if (!closestBySelector(element, '[on*="timeUpdate:"]')) {
```
Comment (aghassemi) [inline extensions/amp-video-service/0.1/amp-video-service.js]: this won't be enough due to chaining support (e.g. `on=tap:foo;timeUpdate:bar"). Dispatching an event every second doesn't seem like a huge perf issue but worth fixing, I just don't think this is the right way. We would want the Event/Action service to do the parsing as it does now and have it expose enough information for cases like this to query what event handlers are installed for a component.

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 9]
```diff
@@ -76,14 +148,159 @@ export class VideoService {
...
+export class VideoEntry {
+
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!VideoService} videoService
+   * @param {!../../../src/video-interface.VideoInterface} video
+   */
+  constructor(ampdoc, videoService, video) {
+
+    /** @private @const{!../../../src/service/ampdoc-impl.AmpDoc} */
```
[the remark this answers is not in the document]
Comment (alanorozco) [reply extensions/amp-video-service/0.1/amp-video-service.js]: Fixed.

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 10]
```diff
@@ -76,14 +148,159 @@ export class VideoService {
...
+    this.ampdoc_ = ampdoc;
+
+    /** @private @const {!VideoService} */
+    this.service_ = videoService;
+
+    /** @private @const{!../../../src/video-interface.VideoInterface} */
```
[the remark this answers is not in the document]
Comment (alanorozco) [reply extensions/amp-video-service/0.1/amp-video-service.js]: Fixed.

File: extensions/amp-video-service/0.1/amp-video-service.js (hunk 4 above)
Comment (alanorozco) [reply extensions/amp-video-service/0.1/amp-video-service.js]: Done.

File: extensions/amp-video-service/0.1/amp-video-service.js (hunk 5 above)
Comment (alanorozco) [reply extensions/amp-video-service/0.1/amp-video-service.js]: Done.

File: extensions/amp-video-service/0.1/amp-video-service.js (hunk 6 above)
Comment (alanorozco) [reply extensions/amp-video-service/0.1/amp-video-service.js]: Done.

File: extensions/amp-video-service/0.1/amp-video-service.js (hunk 7 above)
Comment (alanorozco) [reply extensions/amp-video-service/0.1/amp-video-service.js]: :)

File: extensions/amp-video-service/0.1/amp-video-service.js (hunk 8 above)
Comment (alanorozco) [reply extensions/amp-video-service/0.1/amp-video-service.js]: sg, removed.

File: extensions/amp-video-service/0.1/amp-video-service.js [hunk 11]
```diff
@@ -76,14 +146,160 @@ export class VideoService {
...
+    this.service_.onTick(() => {
+      if (!this.isPlaying_) {
+        return;
+      }
+      handler();
+    });
+  }
+
+  /** */
+  install() {
```
Comment (aghassemi) [inline extensions/amp-video-service/0.1/amp-video-service.js]: nit: can become private now

Comment (alanorozco) [reply extensions/amp-video-service/0.1/amp-video-service.js]: Hmm, I don't think so, since it's called from a static context.

**Pull request merged**